### PR TITLE
scheduler flow control based on current writing kvs count

### PR DESCRIPTION
**Pull request by zhangjinpeng1987**
One write request can associate thousands of kvs, or single pair kv, so flow control based on requests count is irrational.

### CODE REVIEW

File: src/storage/config.rs [hunk 1]
```diff
@@ -23,7 +23,11 @@ const DEFAULT_GC_RATIO_THRESHOLD: f64 = 1.1;
 const DEFAULT_SCHED_CAPACITY: usize = 10240;
 const DEFAULT_SCHED_MSG_PER_TICK: usize = 1024;
 const DEFAULT_SCHED_CONCURRENCY: usize = 102400;
-const DEFAULT_SCHED_TOO_BUSY_THRESHOLD: usize = 4000;
+
+// Assume one TiKV node can write 100_000 kvs per second,
+// and write requests' average process time is about 100ms,
+// so here use 10_000 as the default value.
+const DEFAULT_SCHED_TOO_BUSY_THRESHOLD: usize = 10_000;
```
Comment (huachaohuang) [inline src/storage/config.rs]: Don't quite understand the explanation.

Comment (siddontang) [reply src/storage/config.rs]: Same too.

File: src/storage/txn/scheduler.rs [hunk 2]
```diff
@@ -179,6 +179,7 @@ fn execute_callback(callback: StorageCb, pr: ProcessResult) {
 pub struct RunningCtx {
     cid: u64,
     cmd: Option<Command>,
+    kvs_count: usize,
```
Comment (siddontang) [inline src/storage/txn/scheduler.rs]: @queenypingcap 

Is it better to use kv_count here?

File: src/storage/config.rs (hunk 1 above)
Comment (zhangjinpeng1987) [reply src/storage/config.rs]: You can imaging that the throughput of a pipe equals crosscut area of the pipe multiple the speed of water. In here the speed of water is same as how fast a request can process, the pipe size is same as the threshold above.

Comment (zhangjinpeng1987) [reply src/storage/config.rs]: https://en.wikipedia.org/wiki/Little%27s_law

Comment (siddontang) [reply src/storage/config.rs]: I think the threshold may be a little low if we consider KV count here.  100ms write seems too long.

Comment (zhangjinpeng1987) [reply src/storage/config.rs]: You misunderstood it, the faster request can be handled, the less kv can be piled up.

File: etc/config-template.toml [hunk 3]
```diff
@@ -55,9 +55,9 @@
 # also should less than total cpu cores.
 # scheduler-worker-pool-size = 4
 
-# A write flow control mechanism. When concurrent running write count over
+# A write flow control mechanism. When concurrent running write kvs count over
 # this threshold, will return too busy error.
```
Comment (queenypingcap) [inline etc/config-template.toml]: -> When the number of concurrent running writes exceeds this threshold, the "too busy" (or the exact error message content) error is displayed.

File: src/storage/config.rs [hunk 4]
```diff
@@ -23,7 +23,11 @@ const DEFAULT_GC_RATIO_THRESHOLD: f64 = 1.1;
 const DEFAULT_SCHED_CAPACITY: usize = 10240;
 const DEFAULT_SCHED_MSG_PER_TICK: usize = 1024;
 const DEFAULT_SCHED_CONCURRENCY: usize = 102400;
-const DEFAULT_SCHED_TOO_BUSY_THRESHOLD: usize = 4000;
+
+// Assume one TiKV node can write 100_000 kvs per second,
+// and write requests' average process time is about 100ms,
+// so here use 10_000 as the default value.
```
Comment (queenypingcap) [inline src/storage/config.rs]: Assuming you can write 100_000 KVs per second, and it takes about 100ms to process the write requests on average, hence using the 10_000 as the default value here.

**Comment by siddontang** [comment]
LGTM
PTAL @disksing @huachaohuang

**Pull request merged**